### ISSUE TITLE
Columns sorting - stay on the first result

### ISSUE THREAD

**Issue by Beersteddy** [xavery/transgui#25]
Windows 10, latest transgui as of 14/05/2023

Current behaviour:
- click on any column label to sort ascending/descending `name` for example
- the list will scroll down or up as selection is persistent.
To get to the top of the list need to scroll up through all of the torrents

Proposed behaviour:
Clicking on any column will unselect current torrent and so the first result will be on the top.

Actually quite hard to explain properly.
Basically it should behave like this example list:
https://hdvinnie.github.io/Private-Trackers-Spreadsheet/

**Comment by xavery** [comment]
Thanks for the report. Yeah, I understand what you mean and it makes a lot of sense. It's actually the default behaviour in all other GUIs (i.e. not just torrent clients), I've just gotten so used to this particular quirk in TransGUI that I don't notice it anymore.

**Comment by xavery** [comment]
Unfortunately, it doesn't look like it's possible to have nothing selected in `TVarGrid` if there's at least one row in it, or I can't figure out how to do this. The current code for sorting does preserve the current selection by saving its state in `RowOptions` and then restoring it after the sorting is done. However, LCL code in `TCustomGrid.SetRow` calls `MoveExtend` with `ForceFullyVisible` set to true, which then calls `ScrollToCell`. It doesn't look like this behaviour can be made runtime-dependent without making changes to LCL, so at least for now, setting the row means that the grid will automatically scroll to the selected row.

I disabled behaviour by setting the row to -1 after sorting if an extra property is set. This means that the first row in the table will be automatically selected after sorting, which may not be perfect but is still better than the current behaviour, I think.

The new behaviour is only active for the torrents table, I chose to keep the old behaviour in all other lists as I suspect not everybody will be 100% happy about this change.

You can download a build with this change at https://github.com/xavery/transgui/suites/12912061047/artifacts/697190047 .

**Comment by Beersteddy** [comment]
Thanks for trying to get in to it. Really appreciate.

It works... kind of.
New behaviour works for a while, then after clicking couple of time on `name` `label` etc sorting completely stops to work. Literally does nothing. Manage to repeat the issue twice.... couldn't when I tried 3rd time.
Lets call it success.

`Dzieki`

**Comment by Beersteddy** [comment]
New info.
This just showed up
![image](https://github.com/xavery/transgui/assets/91010735/ab7586e2-5239-4f3d-b3bb-17ad8a310aad)

**Comment by xavery** [comment]
Yeah, it's flaky as hell on Windows, I could see what you're describing as well. Worked fine when I tested it on Linux but that doesn't really help.

I have no other idea on how to approach this right now, sorry.

**Comment by xavery** [comment]
I gave this another shot, will need some more work but looks promising so far. I didn't want to patch LCL code just for TransGUI but it seems like there's no other way and we're building everything from scratch anyway so why not.

@Beersteddy Can you try out this build? https://github.com/xavery/transgui/suites/13024903673/artifacts/705579187

**Comment by Beersteddy** [comment]
@xavery 
Testing it right now. I'll report any problems found.

Previous build wasn't that bad actually. Crashed once a while when left in background, but nothing major really. It was actually a good feature :)
Really appreciate that you're trying to get it fixed and that you continue abandoned project.
Transgui is really outstanding for TM.

BTW.
Do you think is possible to convert  transgui to webUI for transmission - just like combustion?

**Comment by xavery** [comment]
I don't do any web development so I'm not really the right person to ask. Lazarus has something called Pas2JS but I have no idea what it really does.

**Comment by Beersteddy** [comment]
No worries.

So far I did have one single crash when left transgui running in the background for a very long while.
Overall works well enough for what I need.
Hopefully one day someone gonna port it to webui.
![image](https://github.com/xavery/transgui/assets/91010735/b20ed675-db41-4595-b719-0a4afc7da77f)

**Comment by xavery** [comment]
Very glad to hear, as mentioned before this needs some more work but I should be merging it this week if nothing comes up.

**Comment by xavery** [comment]
> No worries.
> 
> So far I did have one single crash when left transgui running in the background for a very long while. Overall works well enough for what I need. Hopefully one day someone gonna port it to webui. ![image](https://user-images.githubusercontent.com/91010735/239733729-b20ed675-db41-4595-b719-0a4afc7da77f.png)

This seems to be https://github.com/transmission-remote-gui/transgui/issues/1459 , so not necessarily related to my changes.

Anyhow, this is now merged by using my own fork of Lazarus/LCL which allows skipping the scrolling.

**Comment by xavery** [comment]
@garoto Thanks a lot. Looks like my gut feeling telling me not to make a new release right away was right this time...

I managed to reproduce the crash without even trying too hard. The only Windows-exclusive change which has been made in the last few weeks (both in my fork and the community project) was https://github.com/xavery/transgui/commit/76d3d6d8c956d03f021434191a11907c3ae1d398 and reverting it seems to make the crash go away.

Please try with the newest master build if you can : https://github.com/xavery/transgui/suites/13068302160/artifacts/708772515

**Comment by garoto** [comment]
@xavery It seems your supposition is indeed correct! The build you linked above is running crash-free while tray-minimized for the past ~25min. Yay?

**Comment by Beersteddy** [comment]
Same. Here. Works just fine, haven't noticed any problems so far.

Thanks xavery

**Comment by blacklion** [comment]
I think fixing this is a big mistake. Current "official" GUI build always left current torrent selected and it is VERY useful in many situations.

For example: you have torrents sorted by size, but want to see torrents with names similar to THIS torrent. You select THIS torrent and sort by name - you see all similar names, as THIS torrent still selected!

To go to top of the list you don't need to "scroll up through all of the torrents" you need to press "Home" button on keyboard.

IMHO, old behavior is much, much better.
